KLR/HLR/AST.lean

```diff
@@ -0,0 +1,278 @@
+  -- reshape a tensor to the specified shape
+  | reshape (a : Var) (shape : Shape)
+  -- broadcast a tensor to the specified shape
+  | broadcast (a : Var) (shape : Shape) (broadcastDims : List Nat)
+  -- create a constant tensor with the given values and shape
+  | const (values : StableHLO.Parsing.DenseLiteral) (shape : Shape) (dtype : TensorLib.Dtype)
+  -- gather elements from a tensor using the provided indices and offset dimensions
+  | gather (input indices : Var) (offsetDims collapsedSliceDims startIndexMap : List Nat) (indexVectorDim : Nat)
+  -- slice a tensor along specified dimensions, with start, limit, and stride
+  | slice (a : Var) (start limit stride : List Nat)
+  -- call another function, passing input values and receiving outputs
+  | call (callee : String) (inputValues : List Var)
+deriving Inhabited, Repr
+
+-- A statement in HLR (High Level Representation).
+-- In SSA form, so each variable is assigned exactly once.
+inductive Statement where
+  -- a comment in the code, for making the dumped IR readable
+  | comment (msg : String)
+  -- assign the result of `op` to `dest` , with the shape `shape`
+  | assign (dest : Var) (op : Operator) (shape : TensorTy)
+  -- return variable `name` from the function
+  | ret (vars : List Var)
+deriving Inhabited, Repr
+
+-- An HLR function. Note that arguments are referred to by index, so
+-- we only store the argument shapes, not names.
+structure Function where
+  name : String
+  inputs : List TensorTy
+  outputs : List TensorTy
+  statements : List Statement
+deriving Inhabited, Repr, Nonempty
+
+-- An HLR program
+structure Program where
+  functions : List Function
+deriving Inhabited, Repr, Nonempty
+
+-- Returns the list of variables that this operator immediately depends on.
+def dependencies : Operator → List Var
+  | .arg _ => []
+  | .binaryOp _ a b => [a, b]
+  | .unaryOp _ a => [a]
+  | .reductionOp _ a b _ => [a, b]
+  | .batchMatmul a b => [a, b]
+  | .arange .. => []
+  | .concat tensors _ => tensors
+  | .select cond a b => [cond, a, b]
+  | .full .. => []
+  | .transpose a _ => [a]
+  | .split_with_sizes a _ => [a]
+  | .reshape a _ => [a]
+  | .broadcast a .. => [a]
+  | .const .. => []
+  | .gather a i .. => [a, i]
+  | .slice a .. => [a]
+  | .call _ inputs => inputs
+
+-- Returns the list of all variables in this function.
+def vars (f : Function) : List Var :=
+  f.statements.filterMap (fun stmt =>
+    match stmt with
+    | .assign dest .. => .some dest
+    | _ => .none)
+
+-- Finds the operator that assigns to a variable in the function.
+def findVar (f : Function) (v : Var) : Option Operator :=
+  f.statements.findSome? (fun stmt =>
+    match stmt with
+    | .assign dest op _ => if dest == v then .some op else .none
+    | _ => .none)
+
+-- Returns the list of all variables that this function depends on transitively.
+def transitiveDependencies (f : Function) (v : Var) : List Var := Id.run do
+  let mut deps := [v]
+  repeat
```

Comment:
Wow, never seen `repeat`. How can this not be marked `partial`?

Reply:
Upon further inspection, no idea why this doesn't need a termination proof. It's moot though, because this code was unused so I removed it.